lib/compat/wordpress-6.0/block-editor-settings.php

```diff
@@ -34,49 +68,39 @@ function_exists( 'gutenberg_is_edit_site_page' ) &&
 	}
 
 	if ( 'other' === $context ) {
+		global $wp_version;
+		$is_wp_5_8 = version_compare( $wp_version, '5.8', '>=' ) && version_compare( $wp_version, '5.9', '<' );
+		$is_wp_5_9 = version_compare( $wp_version, '5.9', '>=' ) && version_compare( $wp_version, '6.0', '<' );
+
 		// Make sure the styles array exists.
 		// In some contexts, like the navigation editor, it doesn't.
 		if ( ! isset( $settings['styles'] ) ) {
 			$settings['styles'] = array();
 		}
 
+		// Remove existing global styles provided by core.
 		$styles_without_existing_global_styles = array();
 		foreach ( $settings['styles'] as $style ) {
 			if (
-				! isset( $style['__unstableType'] ) ||
-				// '__unstableType' is 'globalStyles' for WordPress 5.8 and 'presets' for WordPress 5.9.
-				//
-				// Note that styles classified as'theme', can be from the theme stylesheet
-				// or from the theme.json (the styles section).
-				// We are unable to identify which is which, so we can't remove and recreate those.
-				// Instead, we reload the theme.json styles from the plugin.
-				// Because they'll use the same selectors and load later,
-				// they'll have higher priority than core's.
-				//
-				// Theoretically, this approach with 'theme' styles could be problematic:
-				// if we remove style properties in the plugin, if selectors change, etc.
-				// We need to address this issue directly in core by alowing to identify
-				// styles coming from theme.json.
-				//
-				// A final note about 'theme' styles: this flag is used to identify theme
-				// styles that may need to be removed if the user toggles
-				// "Preferences > Use theme styles" in the preferences modal.
-				//
-				! in_array( $style['__unstableType'], array( 'globalStyles', 'presets' ), true )
+				( $is_wp_5_8 && ! gutenberg_is_global_styles_in_5_8( $style ) ) ||
+				( $is_wp_5_9 && ! gutenberg_is_global_styles_in_5_9( $style ) )
```

Comment:
This PR is only concerned with fixing the bug for 5.9, so it can land when it's ready. If we agree on a direction for https://github.com/WordPress/wordpress-develop/pull/2542 I can also add an extra condition here for 6.0.

Reply:
Let's add the condition and go with this solution, it's the cleanest I think 👍 Thanks for adding these 5.8 and 5.9 fallbacks.